### PR TITLE
COMPUTE-1051: Harden GetSidecarsStatus() against missing container statuses

### PR DESCRIPTION
**Pull request by EricMountain**
Harden GetSidecarsStatus() against missing container statuses

**What type of PR is this?**

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/kubelet/status/status_manager.go [hunk 1]
```diff
@@ -682,34 +682,47 @@ type SidecarsStatus struct {
...
 					}
 				} else if status.State.Waiting != nil {
 					// check if non-sidecars have started
-					klog.Infof("Pod: %s: %s: non-sidecar waiting", format.Pod(pod), container.Name)
+					klog.Infof("Pod: %s: %s: non-sidecar waiting (%s)", format.Pod(pod), container.Name, caller)
 					sidecarsStatus.ContainersWaiting = true
 				}
 			}
 		}
+		if !foundStatus && isSidecar {
```
Comment (nyodas) [inline pkg/kubelet/status/status_manager.go:722]: This could move up and bail early.

File: pkg/kubelet/status/status_manager.go [hunk 2]
```diff
@@ -682,34 +682,47 @@ type SidecarsStatus struct {
 }
 
 // GetSidecarsStatus returns the SidecarsStatus for the given pod
-func GetSidecarsStatus(pod *v1.Pod) SidecarsStatus {
+func GetSidecarsStatus(pod *v1.Pod, caller string) SidecarsStatus {
```
Comment (nyodas) [inline pkg/kubelet/status/status_manager.go:685]: Passing args just for logs seems wrong to me outside a need for debug.
I don't think we should keep those long term.

Comment (EricMountain) [reply pkg/kubelet/status/status_manager.go:685]: Agreed, it's really just for debug. Though honestly given these calls interleave having some kind of trace id is almost vital. I'm in 2 minds about merging the PR with it or not.

File: pkg/kubelet/status/status_manager.go (hunk 1 above)
Comment (EricMountain) [reply pkg/kubelet/status/status_manager.go:729]: I've run through this in my mind a few different ways and nothing stands out that seems good.

I have however added a `break` out of the loop in the `if` above to bail early once a status is found for a given container.

File: pkg/kubelet/status/status_manager.go (hunk 2 above)
Comment (EricMountain) [reply pkg/kubelet/status/status_manager.go:685]: I've removed the caller arg.

File: pkg/kubelet/status/status_manager.go [hunk 3]
```diff
@@ -683,20 +683,35 @@ type SidecarsStatus struct {
...
 	if pod == nil {
 		klog.Infof("Pod was nil, returning empty sidecar status")
 		return SidecarsStatus{}
 	}
-	if pod.Spec.Containers == nil || pod.Status.ContainerStatuses == nil {
-		klog.Infof("Pod Containers or Container status was nil, returning empty sidecar status")
+	if pod.Spec.Containers == nil {
+		klog.Infof("Pod Containers was nil, returning empty sidecar status")
 		return SidecarsStatus{}
 	}
```
Comment (EricMountain) [inline pkg/kubelet/status/status_manager.go:694]: We could consider being even more defensive and returning `sidecarsPresent:true` + `sidecarsReady:false` in these 2 `if`s.

**Review by nyodas — APPROVED** [review]
This looks ok to me.
There might be improvement that could be made by extracting the loop as an external function
And possibly something could also be made if the index was the same.
Preparing the status list beforehand to have the same index would make this faster and more understandable ( no need for a loop )

**Pull request merged**